File: trees/binary_tree_right_side_bfs.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {TreeNode} root
# @return {Integer[]}
def right_side_view(root)
    right_view = []

    return right_view if root.nil?

    queue = [root]

    while !queue.empty?
        right_view << queue.last.val

        level_size = queue.size
        level_size.times do
            current_node = queue.shift
            queue << current_node.left if current_node.left
            queue << current_node.right if current_node.right
        end
    end

    right_view
end

File: trees/binary_tree_right_side_bfs.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        right_view = []

        if root is None:
            return right_view

        queue = deque([root])

        while queue:
            right_view.append(queue[-1].val)

            for _ in range(len(queue)):
                current_node = queue.popleft()
                if current_node.left:
                    queue.append(current_node.left)

                if current_node.right:
                    queue.append(current_node.right)

        return right_view